src/smpp_socket.erl

```diff
@@ -1000,3 +1035,69 @@ check_in_flight(#{in_flight := InFlight, in_flight_limit := Limit} = State)
     {ok, State};
 check_in_flight(State) ->
     {error, {in_flight_limit_cooldown, State, [postpone]}}.
+
+%%%-------------------------------------------------------------------
+%%% ESME back pressure
+%%%-------------------------------------------------------------------
+-define(PT_MAX_AWAIT_REQS(Id), {?MODULE, Id, max_await_reqs}).
+-define(PT_AWAIT_REQS_COUNTER(Id), {?MODULE, Id, await_reqs_counter}).
+
+-spec init_req_counter(state()) -> state().
+init_req_counter(#{max_await_reqs := undefined} = State) ->
+    State;
+init_req_counter(#{max_await_reqs := MaxAwaitReqs, id := Id} = State) ->
+    %% Back pressure here does not rely on exact counter value by design.
+    %% I.e. a bit more than MaxAwaitReqs can be queued due to a natural race
+    %% but then back pressure will kick in anyway. That is why 'write_concurrency'
+    %% counters option fits well here.
+    CounterRef = counters:new(1, [write_concurrency]),
+    ok = persistent_term:put(?PT_AWAIT_REQS_COUNTER(Id), CounterRef),
+    ok = persistent_term:put(?PT_MAX_AWAIT_REQS(Id), MaxAwaitReqs),
+    State#{await_reqs_counter => CounterRef}.
+
+-spec enqueue_req(gen_statem:server_ref()) -> ok | error.
+enqueue_req(Ref) ->
+    enqueue_req(Ref, undefined).
+
+-spec enqueue_req(gen_statem:server_ref(), millisecs() | undefined) -> ok | error.
+enqueue_req(Pid, _Timeout) when is_pid(Pid) ->
```

Comment:
If client does API calls with plain pids instead of registry names -- registry mechanism is outside of responsibility boundaries for this library. Is this even possible to do reverse registry lookup? 

Upd: probably registration of names and/or their respective pids should be doubled in separate module, in order to lookup by two keys for persistent_terms (i.e. counters). But housekeeping of this index will require monitoring too, bcs supervision of `smpp_socket`'s is also outside of responsibility boundaries for this library.

Reply:
Yep, all true. Proper support for direct communication with smpp_socket process via PID is a separate and bigger issue outside of backpressure feature scope. That's why I put a ToDo comment below.